### PR TITLE
Width bucket

### PR DESCRIPTION
**Pull request by cpcloud**
Adds `width_bucket` to make equi-width histograms easier to construct.

- [x] more tests
- [x] ~~should we allow more types?~~

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/math.rst [hunk 1]
```diff
@@ -182,3 +182,8 @@ Floating Point Functions
 .. function:: nan() -> double
 
     Returns the constant representing not-a-number.
+
+.. function:: width_bucket(x, bound1, bound2, n) -> bigint
```
Comment (martint) [inline presto-docs/src/main/sphinx/functions/math.rst]: Use `lower` and `upper` for the names of the bounds

Comment (cpcloud) [inline presto-docs/src/main/sphinx/functions/math.rst]: The standard specifies that the upper and lower bounds can be either the first or second bound. That's why I named them this way. Should I still rename them?

File: presto-docs/src/main/sphinx/functions/math.rst [hunk 2]
```diff
@@ -182,3 +182,8 @@ Floating Point Functions
...
+
+    Returns the bin number of ``x`` in an equidepth histogram with the
```
Comment (martint) [inline presto-docs/src/main/sphinx/functions/math.rst]: The ANSI SQL spec defines this function as producing an *equi-width* histogram:

> `<width bucket function>` is a function of four arguments, returning an integer between 0 (zero) and the value of the  final argument plus 1 (one), by assigning the  first argument to an equi-width partitioning of the range of numbers between the second and third arguments. Values outside the range between the second and third arguments are assigned to either 0 (zero) or the value of the final argument plus 1 (one).

File: presto-docs/src/main/sphinx/functions/math.rst (hunk 1 above)
Comment (martint) [inline presto-docs/src/main/sphinx/functions/math.rst]: Ah, you're right (just took a look at the spec). Leave them as they are.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 3]
```diff
@@ -392,4 +392,43 @@ private static void checkRadix(long radix)
...
+        checkCondition(bound1 != bound2, INVALID_FUNCTION_ARGUMENT, "bounds cannot equal each other");
+
+        long result = 0;
+
+        if (bound1 < bound2) {
+            if (operand < bound1) {
+                result = 0;
+            }
+            else if (operand >= bound2) {
+                result = Math.addExact(bucketCount, 1);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Catch `ArithmeticException` and throw a `PrestoException` with `NUMERIC_VALUE_OUT_OF_RANGE` error code.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 4]
```diff
@@ -392,4 +392,43 @@ private static void checkRadix(long radix)
...
+            else {
+                result = (long) ((double) bucketCount * (operand - bound1) / (bound2 - bound1) + 1);
+            }
+        }
+        else if (bound1 > bound2) {
+            if (operand > bound1) {
+                result = 0;
+            }
+            else if (operand <= bound2) {
+                result = Math.addExact(bucketCount, 1);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Same as above

File: presto-docs/src/main/sphinx/functions/math.rst (hunk 2 above)
Comment (cpcloud) [inline presto-docs/src/main/sphinx/functions/math.rst]: Whoops meant to write equi width there. I honk I was reading something about equi depth and had a brain slip

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 5]
```diff
@@ -392,4 +392,43 @@ private static void checkRadix(long radix)
...
+                result = 0;
+            }
+            else if (operand >= bound2) {
+                result = Math.addExact(bucketCount, 1);
+            }
+            else {
+                result = (long) ((double) bucketCount * (operand - bound1) / (bound2 - bound1) + 1);
+            }
+        }
+        else if (bound1 > bound2) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: This branch is equivalent to the one above with the bounds flipped, no? If so, I'd write the code as:

```java
long lower = Math.min(bound1, bound2);
long upper = Math.max(bound1, bound2);

if (operand < lower) {
    result = 0;
}
else if (operand >= upper) {
    result = Math.addExact(bucketCount, 1);
}
else {
    result = (long) ((double) bucketCount * (operand - lower) / (upper - lower) + 1);
}
```

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 6]
```diff
@@ -519,4 +519,21 @@ public void testFromBase()
         assertInvalidFunction("from_base('Z', 35)", BIGINT, "Not a valid base-35 number: Z");
         assertInvalidFunction("from_base('9223372036854775808', 10)", BIGINT, "Not a valid base-10 number: 9223372036854775808");
     }
+
+    @Test
+    public void testWidthBucket()
+    {
+        assertFunction("width_bucket(3.14, 0, 4, 3)", BIGINT, 3);
+        assertFunction("width_bucket(infinity(), 0, 4, 3)", BIGINT, 4);
+
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: Add a test for operand < lowest bound

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 5 above)
Comment (cpcloud) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Nice.

Comment (cpcloud) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: The branches are slightly different. For example if `bound1 = 0` and `bound2 = 4` then bin 1 starts at 0, whereas if `bound1 = 4` and `bound2 = 0`, then bin 1 starts at 4.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: I see, it reverses the order of the buckets. So, if I'm not mistaken, it should be sufficient to do this before returning:

```java
if (bound1 > bound2) {
    result = bucketCount - result;
}
```

**Comment by cpcloud** [comment]
Should I implement this for `ARRAY<DOUBLE>`, where the array would specify the bin edges?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 7]
```diff
@@ -392,4 +392,42 @@ private static void checkRadix(long radix)
...
+
+        if (operand < lower) {
+            result = 0;
+        }
+        else if (operand >= upper) {
+            try {
+                result = Math.addExact(bucketCount, 1);
+            }
+            catch (ArithmeticException e) {
+                throw new PrestoException(NUMERIC_VALUE_OUT_OF_RANGE, "bucketCount overflow");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: I'd rephrase as "bucket for value xxx is out of range"

**Comment by martint** [comment]
> Should I implement this for ARRAY<DOUBLE>, where the array would specify the bin edges?

That makes sense, but I would do it as a separate PR or commit.

**Comment by cpcloud** [comment]
I think `result = bucketCount - result` should actually be `result = (bucketCount - result) + 1` since our buckets indices are 1-based.

**Comment by martint** [comment]
> I think result = bucketCount - result should actually be result = (bucketCount - result) + 1 since our buckets indices are 1-based.

Yes, you're right.

**Comment by martint** [comment]
> > I think result = bucketCount - result should actually be result = (bucketCount - result) + 1 since our buckets indices are 1-based.

> Yes, you're right.

Make sure there's a test that catches that issue.

**Comment by cpcloud** [comment]
> Make sure there's a test that catches that issue.

An existing test caught it :)

**Comment by martint** [comment]
Merged, thanks!
